File: 206-Reverse-Linked-List/v1.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next



class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.recursiveReverseList(head, None)

    def recursiveReverseList(self, head: Optional[ListNode], prev: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            # Base case
            return prev
        else:
            # Step case
            return self.recursiveReverseList(head.next, ListNode(head.val, prev))

File: 143-Reorder-List/v1.java
class Solution {
    public void reorderList(ListNode head) {
        if (head.next == null) {
            return;
        } 
        if (head.next.next == null) {
            return;
        }

        ListNode lastButOne = head;

        while (lastButOne.next.next != null) {
            lastButOne = lastButOne.next;
        }

        ListNode last = lastButOne.next;
        lastButOne.next = null;

        reorderList(head.next);
        last.next = head.next;
        head.next = last;

        return;
    }
}
